### ISSUE TITLE
Investigate switching to ArrayBufferWriter impl to simplify grow logic when reading from stream in JsonSerializer

### ISSUE THREAD

**Issue by ahsonkhan** [dotnet/runtime#32355]
https://github.com/dotnet/runtime/blob/d4b06b1a9e3c56e343c42efad2211c04c196a0cf/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs#L95

cc @steveharter

**Comment by davidfowl** [comment]
ArrayBufferWriter doesn't pool currently. That's likely a performance regression

**Comment by ahsonkhan** [comment]
Yep, agreed. Since this is used internally, in this case, we can consider using `PooledByteBufferWriter` just like what we do on the Serialize/Write side.

**Comment by scalablecory** [comment]
Consider using the `ArrayBuffer` we use in System.Net.

**Comment by ahsonkhan** [comment]
That's essentially identical to `PooledByteBufferWriter`, if it also implemented the `IBufferWriter<byte>` interface. We should consider merging these and having only one copy.

https://github.com/dotnet/runtime/blob/f06f46c049e442d34950884783b4b85f7e6c1532/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/PooledByteBufferWriter.cs#L13-L16

**Comment by ahsonkhan** [comment]
The source still has the todo comment, likely worth addressing manually, or closing explicitly.
